downloader.go

```diff
@@ -174,42 +176,87 @@ func (dl *Downloader) listAvailableFiles(audioLibrary *AudioFileLibrary, referen
-	// Just return true for now.
+// Check that the sound file is valid.
+func (dl *Downloader) validateSoundFile(fileNameOnDisk string, fileSize int) bool {
+
+	// Check size on disk is the same as the size the api-server tells us this file should be.
+	file, err := os.Open(fileNameOnDisk)
+	if err != nil {
+		return false
+	}
+
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return false
+	}
+
+	if fileInfo.Size() != int64(fileSize) {
+		return false
+	}
+
 	return true
+
 }
 
-func (dl *Downloader) downloadAllNewFiles(audioLibrary *AudioFileLibrary, referencedFiles []int) {
-	log.Println("Starting downloading audio files.")
-	for _, fileID := range referencedFiles {
-		if _, exists := audioLibrary.GetFileNameOnDisk(fileID); exists {
-			valid := dl.validateSoundFile(audioLibrary, fileID)
-			if exists && valid {
-				continue
-			}
-		} else {
-			log.Printf("Attempting to download file with id %d", fileID)
+// Removes a file from the disk of the device.  Also takes it out of the audioLibrary so it won't be accessed later.
+func (dl *Downloader) removeAudioFile(audioLibrary *AudioFileLibrary, fileID int, fileNameOnDisk string) {
 
-			fileInfo, err := dl.api.GetFileDetails(fileID)
-			if err != nil {
-				log.Printf("Could not download file with id %d. Downloading next file", fileID)
-				continue
-			}
+	delete(audioLibrary.FilesByID, fileID)
 
-			fileNameOnDisk := MakeFileName(fileInfo.File.Details.OriginalName, fileInfo.File.Details.Name, fileID)
+	err := os.Remove(fileNameOnDisk)
+	if err != nil {
+		log.Printf("Could not remove file with ID %d and name %s from disk. Error is: %s", fileID, fileNameOnDisk, err)
+	}
+
+}
 
-			if err = dl.api.DownloadFile(fileInfo, filepath.Join(dl.audioDir, fileNameOnDisk)); err != nil {
-				log.Printf("Could not download file with id %d.  Error is %s. Downloading next file", fileID, err)
-			} else {
-				// Add this file to our audio library.
+// Try and download a single audio file from the API server.
+func (dl *Downloader) downloadAudioFile(audioLibrary *AudioFileLibrary, fileID int, fileResp *api.FileResponse) {
+
+	fileNameOnDisk := MakeFileName(fileResp.File.Details.OriginalName, fileResp.File.Details.Name, fileID)
+
+	log.Printf("Processing file with id %d and name %s", fileID, fileNameOnDisk)
+
+	for i := 1; i <= maxDownloadRetries; i++ {
+		if err := dl.api.DownloadFile(fileResp, filepath.Join(dl.audioDir, fileNameOnDisk)); err != nil {
+			log.Printf("Error dowloading sound file id %d and name %s.  Error is %s.", fileID, fileNameOnDisk, err)
+		} else {
+			if dl.validateSoundFile(filepath.Join(dl.audioDir, fileNameOnDisk), fileResp.FileSize) {
+				// File is valid, add it to our audio library.
 				audioLibrary.FilesByID[fileID] = fileNameOnDisk
+				return
 			}
+			log.Printf("File with ID %d and name %s is not valid. Removing from disk.", fileID, fileNameOnDisk)
+			dl.removeAudioFile(audioLibrary, fileID, filepath.Join(dl.audioDir, fileNameOnDisk))
+		}
+		if i < maxRetries {
```

Comment:
The loop is checking `maxDownloadRetries` while this is checking `maxRetries`. Smells like a bug to me.

What's the difference between these consts?

A less bug prone way to structure this loop is {
```go

int i
for {
    // attempt download, return on success (as per existing code)

    if i < maxRetries {
        log.Println("Trying again in", retryDownloadInterval)
        time.Sleep(retryDownloadInterval)
        i++
    } else {
        break
    }
}

log.Printf("Could not download and validate file with ID %d and name %s.", fileID, fileNameOnDisk)
```

That way the management of the loop's state is all in one place.